pfb/wavelets/convolution.py

```diff
@@ -0,0 +1,246 @@
+import numba
+
+from pfb.wavelets.modes import Modes
+
+
+@numba.njit(nogil=True, cache=True)
+def upsampling_convolution_valid_sf(input, filter, output, mode):
+    if mode is Modes.periodisation:
+        raise NotImplementedError("periodisation upsampling not implemented")
+
+    N = input.shape[0]
+    F = filter.shape[0]
+    O = output.shape[0]
+
+    if F % 2:
+        raise ValueError("even filter required for upsampling")
+
+    if N < (F // 2):
+        raise ValueError("input.shape[0] < (filter.shape[0] // 2)")
+
+    o = 0
+    i = (F // 2) - 1
+
+    while i < N:
+        sum_even = input.dtype.type(0)
+        sum_odd = input.dtype.type(0)
+        j = 0
+
+        while j < F // 2:
+            sum_even += filter[j*2] * input[i-j]
+            sum_odd += filter[j*2+1] * input[i-j]
+
+            j += 1
+
+        output[o] += sum_even
+        output[o+1] += sum_odd
+
+        i += 1
+        o += 2
+
+
+@numba.njit(nogil=True, cache=True)
+def downsampling_convolution(input, output, filter, mode, step):
+    if mode is Modes.periodisation:
+        raise NotImplementedError("periodisation downsampling not implemented")
```

Comment:
Periodisation is actually handled here and should defer to a function specifically for it.